crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+    S8,
+    /// The type is an unsigned 8-bit integer.
+    U8,
+    /// The type is a signed 16-bit integer.
+    S16,
+    /// The type is an unsigned 16-bit integer.
+    U16,
+    /// The type is a signed 32-bit integer.
+    S32,
+    /// The type is an unsigned 32-bit integer.
+    U32,
+    /// The type is a signed 64-bit integer.
+    S64,
+    /// The type is an unsigned 64-bit integer.
+    U64,
+    /// The type is a 32-bit floating point number.
+    F32,
+    /// The type is a 64-bit floating point number.
+    F64,
+    /// The type is a Unicode character.
+    Char,
+    /// The type is a string.
+    String,
+    /// The type is the unit type.
+    Unit,
+    /// The type is a compound interface type.
+    ///
+    /// The value is a type index to a compound type.
+    Compound(u32),
+}
+
+impl InterfaceType {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Bool => bytes.push(INTERFACE_TYPE_BOOL),
+            Self::S8 => bytes.push(INTERFACE_TYPE_S8),
+            Self::U8 => bytes.push(INTERFACE_TYPE_U8),
+            Self::S16 => bytes.push(INTERFACE_TYPE_S16),
+            Self::U16 => bytes.push(INTERFACE_TYPE_U16),
+            Self::S32 => bytes.push(INTERFACE_TYPE_S32),
+            Self::U32 => bytes.push(INTERFACE_TYPE_U32),
+            Self::S64 => bytes.push(INTERFACE_TYPE_S64),
+            Self::U64 => bytes.push(INTERFACE_TYPE_U64),
+            Self::F32 => bytes.push(INTERFACE_TYPE_F32),
+            Self::F64 => bytes.push(INTERFACE_TYPE_F64),
+            Self::Char => bytes.push(INTERFACE_TYPE_CHAR),
+            Self::String => bytes.push(INTERFACE_TYPE_STRING),
+            Self::Unit => bytes.push(INTERFACE_TYPE_UNIT),
+            Self::Compound(index) => bytes.extend(encoders::u32(*index)),
+        }
+    }
+}
+
+/// Represents a compound interface type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CompoundInterfaceType<'a> {
+    /// A list interface type.
+    List(InterfaceType),
+    /// A record interface type.
+    Record(&'a [(&'a str, InterfaceType)]),
+    /// A variant interface type.
+    Variant(&'a [(&'a str, InterfaceType)]),
+    /// A tuple interface type.
+    Tuple(&'a [InterfaceType]),
+    /// A flags interface type.
+    Flags(&'a [&'a str]),
+    /// An enumeration interface type.
+    Enum(&'a [&'a str]),
+    /// A union interface type.
+    Union(&'a [InterfaceType]),
+    /// An optional interface type.
+    Optional(InterfaceType),
+    /// An expected interface type.
+    ///
+    /// If `error` is `None`, then the error type is unit.
+    Expected {
+        /// The type representing success.
+        ok: InterfaceType,
+        /// The type representing failure.
+        error: Option<InterfaceType>,
+    },
+    /// A named interface type.
+    ///
+    /// Named types are either part of a type's declaration (i.e. a record)
+    /// or may be a type alias.
+    Named(NamedInterfaceType<'a>),
```

Comment:
I'm a little confused what a named variant is here in this case? Is this something like the `name` section but baked directly into the type section?

Reply:
That's correct. This can be used to both define a type alias and also give a "name" to a type (e.g. naming a record type).

It isn't strictly *required* to give types a name as it won't effect the runtime behavior of the component, but the intention behind including it here in the types section rather than some other section is to really promote it as something that's not really optional for a good developer experience when consuming the component (bindings generators can always derive a name of some sort, but _using_ the binding won't be pleasant).

Reply:
Resolving this one so we can continue the "names encoded in the types section" discussion in the other comment.